plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/purity/PurityChecker.kt

```diff
@@ -0,0 +1,180 @@
+package org.jetbrains.kotlin.formver.purity
+
+import org.jetbrains.kotlin.formver.embeddings.callables.UserFunctionEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.*
+import org.jetbrains.kotlin.formver.viper.MangledName
+import org.jetbrains.kotlin.formver.viper.mangled
+
+class PurityChecker() {
+
+    private val registry = mutableMapOf<MangledName, PurityStatus>()
+    private val declaredPure = mutableMapOf<MangledName, Boolean>()
+
+    private val waiting = mutableMapOf<MangledName, MutableSet<UserFunctionEmbedding>>()
+    private var currentEvaluatedFunction: UserFunctionEmbedding? = null
```

Comment:
Mutable state like this should generally be avoided; you probably want the purity checker to be passed into something that works with the function.  (Though I'm not sure we really need to track purity somewhere other than the ProgramConverter?)